[#68374, #68375] Unsaved changes are lost when reordering sections and interacting with outcomes

mrmir opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->
https://community.openproject.org/work_packages/68374
https://community.openproject.org/work_packages/68375

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

mrmir committed:
Add confirmation for section reordering

mrmir committed:
Add confirmation for outcome actions

cbliard reviewed (changes requested):
I found a little issue: it's possible to have the "You have unsaved changes" message when nothing is being edited.

Here is a reproduction screen recording:

[no edit but asks if should discard changes.webm](https://github.com/user-attachments/assets/aa9af0f2-2b78-49fa-9e3b-e0e325f0e626)

And here are the steps:
- edit an item
- start typing in a field
- move another item or a section
- the "Do you want to discard changes" alert is opened, choose yes
  - the edit is discarded and replaced
- move another item or section again, or add / edit an outcome
- 🔴 the "Do you want to discard changes" alert is opened but there is no edit in progress
- 🔴 there is also the "This page has been updated by someone else. Reload to view changes." message being displayed, but I was the one editing it.

And something I'm not sure about: when editing an outcome, and then doing something that would discard it, there is no alert dialog asking if we are sure we want to discard the outcome edit. Is it ok?

And also, when there is no change in the edit, it still asks if we want to discard the changes. Is it ok? (it could avoid asking if nothing was changed).

cbliard reviewed (commented):
Also, removing outcome asks "Are you sure?" first, then after validating it asks "Are you sure you want to discard unsaved changes?". That's good.

But when removing an agenda item, it only asks "Are you sure?". It does not ask about discarding unsaved changes. Is it ok?

mrmir committed:
Reset pageState explicitly when navigating out

This is necessary because the turbo:load, turbo:render, turbo:submit
events that normally reset this are not fired when reloading a component
via turbo streams, and there isn't an equivalent turbo:stream-render
event that can be used instead.

mrmir commented:
Thanks for taking a good look at this @cbliard. Could you check again please?

> I found a little issue: it's possible to have the "You have unsaved changes" message when nothing is being edited.

This should be fixed now with my latest changes. It existed before as well apparently, but with more use of the confirm dialogs now it stood out more. Good eye 🦅 

> And something I'm not sure about: when editing an outcome, and then doing something that would discard it, there is no alert dialog asking if we are sure we want to discard the outcome edit. Is it ok?

> And also, when there is no change in the edit, it still asks if we want to discard the changes. Is it ok? (it could avoid asking if nothing was changed).

These are related issues. When editing an outcome and changing the editor contents, I see the confirm dialogs. It is only when the strings are the same that confirm dialog isn't displayed. I think that makes sense as well, and so I'll create a [follow up WP](https://community.openproject.org/wp/68654) to harmonize that for the agenda item changes as well. The reason the confirm dialog always shows up there is because of the title field, which is only checking for emptiness.

> Also, removing outcome asks "Are you sure?" first, then after validating it asks "Are you sure you want to discard unsaved changes?". That's good.

>But when removing an agenda item, it only asks "Are you sure?". It does not ask about discarding unsaved changes. Is it ok?

Yes, this is because the outcome actions reload the entire list of agenda items, but adding/editing/deleting each agenda item only reloads that particular item (allowing other unsaved changes in different agenda items to be kept). This is a known inconsistency at this point and there are plans to also have the whole list reload for agenda item actions, but not immediately.  I hope that makes sense 😅

mrmir committed:
Merge branch 'release/16.6' into bug/68374-unsaved-changes-are-lost-when-sections-are-reordered

cbliard commented:
> > And something I'm not sure about: when editing an outcome, and then doing something that would discard it, there is no alert dialog asking if we are sure we want to discard the outcome edit. Is it ok?
> 
> These are related issues. When editing an outcome and changing the editor contents, I see the confirm dialogs. It is only when the strings are the same that confirm dialog isn't displayed. I think that makes sense as well, and so I'll create a [follow up WP](https://community.openproject.org/wp/68654) to harmonize that for the agenda item changes as well. The reason the confirm dialog always shows up there is because of the title field, which is only checking for emptiness.

Just to confirm we are talking about the same thing here: I mean when editing an outcome, then when editing an item, it discards the outcome without any confirmation.

Here is a screen recording:

[editing item discards outcome.webm](https://github.com/user-attachments/assets/27854d97-b16e-49c6-8c35-92c590019f68)

Is it expected?

cbliard reviewed (approved):
Looks good to me now 👏

mrmir committed:
Add confirmation for remaining agenda item actions

mrmir commented:
Thanks for that video @cbliard. I was indeed talking about a different case, but that's a good catch. I added the checks in for 'Edit agenda item' and 'Add notes'.

Doing so however creates a different issue @oliverguenther, where now if you're editing an agenda item and select 'Edit agenda item' for a different one, the confirm dialog will show up even though no data loss occurs as two agenda items can be edited simultaneously still. Is that acceptable?

mrmir committed:
Skip flickering spec

cbliard commented:
Too bad for that flickering spec. I was not able to reproduce it either.
